### ISSUE TITLE
[jvm-packages] How to evaluate the model in Spark version

### ISSUE THREAD

**Issue by yxzf** [dmlc/xgboost#1406]
I run the demo of spark version, but I do not know how to evaluate the model. I got the predictions
```
val rawPreds = xgboostModel.predict(new DMatrix(testData))
```

But the rawPreds have no info of label, how can I evaluate? I think there is no guarantee for the order of predictions and labels

**Comment by randomjohnnyh** [comment]
I solved this by copying the predict code under:
https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala

By zipping iterators, you can map to tuples of (predicted, actual label), or you can map directly to error.

**Comment by yxzf** [comment]
@randomjohnnyh Sorry I still do not know how to do, can you paste your code?

**Comment by randomjohnnyh** [comment]
See these two lines:
```
val dMatrix = new DMatrix(new JDMatrix(testSamples, cacheFileName))
Iterator(broadcastBooster.value.predict(dMatrix))
```
Use `dMatrix.getLabel()` to retrieve a float[] of labels. Then `broadcastBooster.value.predict(dMatrix)` gives a float[][] column vector of predictions.

So we can write a new function that is the same except with:
```
val dMatrix = new DMatrix(new JDMatrix(testSamples, cacheFileName))
val labels = dMatrix.getLabel()
val preds = broadcastBooster.value.predict(dMatrix)
Iterator((labels zip preds).map{case (x, y) => x - y(0)}.iterator
```

**Comment by CodingCat** [comment]
you can write an evaluation function, like https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-example/src/main/java/ml/dmlc/xgboost4j/java/example/CustomObjective.java#L104


and 

evaluate by https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-example/src/main/java/ml/dmlc/xgboost4j/java/example/CustomObjective.java#L104

**Comment by yxzf** [comment]
@CodingCat I have two questions
(1) the predict function of XGBoost
Actually, the predict function of XGBoost give the raw prediction, but many others library I have used will give the probability(Binary classification)
I think one parameter in predict in ml.dmlc.xgboost4j.scala.spark.XGBoostModel will be better
  ```
  def predict(testSet: DMatrix): Array[Array[Float]] = {
    _booster.predict(testSet, true, 0)
  }
  ```
  This will give the raw prediction, as in the c++ code, 
  ```
  if (pred_leaf) {
      gbm_->PredictLeaf(data, out_preds, ntree_limit);
    } else {
      this->PredictRaw(data, out_preds, ntree_limit);
      if (!output_margin) {
        obj_->PredTransform(out_preds);
      }
    }
  ```
(2) predict function only gives the prediction value, but loses the label. Although DMatrix.label can give the label, I think no guarantee about the order. 
Johnny Ho's method is OK, and  I think source code includes this will be fine.

**Comment by yxzf** [comment]
@randomjohnnyh Hi, I find a way to predict with label
```
def predict(model:XGBoostModel,testData:RDD[(LabeledPoint)]) : RDD[(Double, Double)] = {

    import DataUtils._
    val booster = model.asInstanceOf[XGBoostModel].booster
    val broadcastBooster = testData.sparkContext.broadcast(booster)

    val result = testData.map { line =>
      val result_array = broadcastBooster.value.predict(new DMatrix(Iterator(line.features)))(0)
      val pred = if (result_array(0) > 0.5) 1.0 else 0.0
      (pred.toDouble, line.label.toDouble)
    }
    result
  }
```
But this way is to iterator every sample and is slow.

I try to change the source code by your suggestion, but I do not how to pass the parameter. The parameter needs mllib.linalg.Vector, how can it get the label if I do not pass the LabelPoint?

```
def predictWithLabel(testSet: RDD[Vector], cache: Boolean = false): RDD[Array[(Float, Float)]] = {
    import DataUtils._
    val broadcastBooster = testSet.sparkContext.broadcast(_booster)
    val appName = testSet.context.appName
    testSet.mapPartitions { testSamples =>
      if (testSamples.hasNext) {
        val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
        Rabit.init(rabitEnv.asJava)
        val cacheFileName = {
          if (cache) {
            s"$appName-dtest_cache-${TaskContext.getPartitionId()}"
          } else {
            null
          }
        }
        val dMatrix = new DMatrix(new JDMatrix(testSamples, cacheFileName))
        val labels = dMatrix.getLabel
        val preds = broadcastBooster.value.predict(dMatrix)
        Rabit.shutdown()
        Iterator((labels zip preds).map{case (x, y) => (x, y(0))})

      } else {
        Iterator()
      }
    }
  }
```

**Comment by peperxuhui** [comment]
I have the same question just as @yxzf said,does anyone has any idea?

**Comment by CodingCat** [comment]
@yxzf what do you mean by no guarantee about order?

**Comment by peperxuhui** [comment]
After using XGBoostModel.eval which in  https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala .
this problem has solved. Thanks!

**Comment by peperxuhui** [comment]
This is My auc eval:
class AucEval extends EvalTrait {
  private val logger: Log = LogFactory.getLog(classOf[AucEval])

  override def getMetric: String = {
    "auc"
  }

  override def eval(predicts: Array[Array[Float]], dmat: DMatrix): Float = {
    var roc: Float = 0
    val label = dmat.getLabel
    var indices = ArrayBuffer[Int]()
    for (i <- predicts.indices) {
      indices += i
    }
    indices = indices.sortWith((s, t) => predicts(s)(0) > predicts(t)(0))
    var tp = 0
    var fp = 0
    for (i <- predicts.indices) {
      if (label(indices(i)) == 1) {
        tp += 1
      }
      else if (label(indices(i)) == 0) {
        roc += tp
        fp += 1
      }
    }
    if (tp == 0 || fp == 0) {
      logger.error("warning: Too few positive true labels or negative true labels\n")
      roc = 0
    }
    else {
      roc = roc / tp / fp
    }
    roc
  }
}

By Using AUC, it just like 
val auc = new AucEval
xgboostModel.eval(testSet, auc, "auc")

**Comment by dirceusemighini** [comment]
@CodingCat I think that what @yxzf want to say, is that RDD doesn't guarantee the order of it's data, so if we use a predict(RDD[Vector]) we don't know the label of each prediction.
It would be better to have a predict method that receives an RDD[LabeledPoint] and returns a tuple RDD[(Double,Double)] where the first element of the tuple would be the label of the LabeledPoint, and the second one, the prediction for that label.
This is the standard in the Spark MLLib.
I tried to do something similar to what @yxzf has done, my code is [here](http://stackoverflow.com/a/39236628/1791289), but it's very slow to execute when I have more than 1 worker.
Even with one worker, I wouldn't be using all of my cluster potential.
Does anybody has a solution for this?

**Comment by CodingCat** [comment]
@dirceusemighini if you only need the preserved order for evaluation, you can directly call eval() method in XGBoostModel

**Comment by yxzf** [comment]
@dirceusemighini I read your code, I do not know why it is slow, but maybe you can try this
```
def labelPredict(testSet: RDD[XGBLabeledPoint],
               useExternalCache: Boolean = false,
               booster: XGBoostModel): RDD[(Float, Float)] = {
val broadcastBooster = testSet.sparkContext.broadcast(booster)
testSet.mapPartitions { testData =>
	val dMatrix = new DMatrix(testData)
	val predictions = broadcastBooster.value.predict(dMatrix).map(p => p(0)).toList
  	val labels = dMatrix.getLabel.toList
  	val results = labels.zip(predictions)
  	results.iterator
}
}
```

**Comment by yxzf** [comment]
@CodingCat Eval returns the evaluation results, but we cannot get the pair<label, prediction>, sometimes we want to get this. Maybe you can give a interface for users, then users can do eval or something they want. And many ml packages, like sklearn, mllib, give this type interface, so many users ask this question.

**Comment by yxzf** [comment]
@dirceusemighini Maybe it is slow is that you do not follow the source code in XGBoostModel.scala, like this. 
```
def predict(testSet: RDD[Vector], useExternalCache: Boolean = false): RDD[Array[Array[Float]]] = {
    import DataUtils._
    val broadcastBooster = testSet.sparkContext.broadcast(_booster)
    val appName = testSet.context.appName
    testSet.mapPartitions { testSamples =>
      if (testSamples.hasNext) {
        val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
        Rabit.init(rabitEnv.asJava)
        val cacheFileName = {
          if (useExternalCache) {
            s"$appName-dtest_cache-${TaskContext.getPartitionId()}"
          } else {
            null
          }
        }
        val dMatrix = new DMatrix(new JDMatrix(testSamples, cacheFileName))
        val res = broadcastBooster.value.predict(dMatrix)
        Rabit.shutdown()
        Iterator(res)
      } else {
        Iterator()
      }
    }
  }
```

**Comment by dirceusemighini** [comment]
@yxzf just updated my code:
def labelPredict(testSet: RDD[XGBLabeledPoint],
                   useExternalCache: Boolean = false,
                   booster: XGBoostModel): RDD[(Float, Float)] = {
    import ml.dmlc.xgboost4j.scala.spark.DataUtils._
    val broadcastBooster = testSet.sparkContext.broadcast(booster)
    val appName = testSet.context.appName
    testSet.mapPartitions { testSamples =>
      if (testSamples.hasNext) {
        val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
        Rabit.init(rabitEnv.asJava)
        val cacheFileName = {
          if (useExternalCache) {
            s"$appName-dtest_cache-${TaskContext.getPartitionId()}"
          } else {
            null
          }
        }
        val dMatrix = new DMatrix(testSamples, cacheFileName)
        val res = broadcastBooster.value.predict(dMatrix).flatten
        val labels = dMatrix.getLabel
        Rabit.shutdown()
        (labels zip res).toIterator
      } else {
        Iterator()
      }
    }
  }
1 worker 85s (before it was 162) local workstation test
20 workers 15min and it hasn't completed until now
This is the test log:
Tracker started, with env={DMLC_NUM_SERVER=0, DMLC_TRACKER_URI=10.42.10.252, DMLC_TRACKER_PORT=9091, DMLC_NUM_WORKER=20}
[Stage 6:>                                                         (0 + 4) / 20]

**Comment by yxzf** [comment]
@dirceusemighini I do not know, maybe coding cat can answer your question...

**Comment by dirceusemighini** [comment]
@yxzf Thanks, for the help, my problem is in training I think that it's better to look, or open another issue

**Comment by CodingCat** [comment]
Hi, @dirceusemighini  and @yxzf , I finally get a chance to look at your discussions carefully, 

First, I'm not sure if I misunderstand something among those words, I think predict() methods in MLLIB returns RDD[Double] instead of RDD[(Double, Double)] as you said, http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.mllib.tree.model.DecisionTreeModel

Second, with predict method `def predict(testSet: RDD[Vector], useExternalCache: Boolean = false): RDD[Array[Array[Float]]] `, it does guarantee the consistent order of the prediction results and the elements in `testSet`, as it is based on `mapPartition `

Third, Typing the parameter as RDD[Vector] instance of RDD[LabeledPoint] is to be consistent with MLLIB  (as you can see from the above link)

Forth, if you have a testset of RDD[LabeledPoint] you can simply get RDD[Vector] by testset.map(_.features), and after you get the prediction results, you can use zip method (http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.rdd.RDD ) to get a RDD of RDD[(LabeledPoint, predictionResults)]

**Comment by yxzf** [comment]
@CodingCat Thank you for your relay. I think you are right. I need learn spark more deeply.

**Comment by dirceusemighini** [comment]
@CodingCat thank you. I've seen this problem in some issues around the project, I think that it would be great to provide an example with this code, I can do that if you are OK with that.

I messed up the methods predict(features: RDD[Vector]) with predict(features: Vector), if we have this second one we can do 
`rdd.map(labeledPoint => (labeledPoint.label,model.predict(labeledPoint.values))`
in just one step, instead of
`rdd.map(_.label).zip(model.predict(rdd.map(_.values)))`
But I don't know if a predict(features: Vector) method fit in XGBoost.

**Comment by yxzf** [comment]
@dirceusemighini Oh, xgboost is slow for predicting one instance one time, see the issue I proposed https://github.com/dmlc/xgboost/issues/1538

**Comment by CodingCat** [comment]
@dirceusemighini sure, you are welcome to provide an example about the code 

and also, as @yxzf points out, XGBoost does not support online prediction for now

**Comment by jerrymouse333** [comment]
Hi @CodingCat 

I'm puzzled about the design of the predict() function. Take the following code as an example:
val trainRDD = sc.parallelize(Seq(
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(1.0, new DenseVector(Array(2.0, 3.0, 4.0))), 
  LabeledPoint(0.0, new DenseVector(Array(5.0, 5.0, 5.0)))
), 4)
val xgboostModel = XGBoost.train(trainRDD, paramMap, 1, 4, useExternalMemory=true)
val xgboostPrediction = xgboostModel.predict(trainRDD.map{x => x.features})
**trainRDD.count()
xgboostPrediction.count()**

## Question 1:
The count of the prediction output: xgboostPrediction is only 4, which is equal to the number of partitions. However, the count of training data: trainRDD is 13. One would naturally expect the count of prediction output to be 13 as well. What is the incentive of this design that the predict() function takes in RDD[vector] and outputs RDD[Array[Array[Double]]]? 

## Question 2:
I can do something like this to change the prediction output to be RDD[Double] using this code: xgboostPrediction.flatMap(x=>x).map(_(0))
However, I'm not 100% sure that it is always safe to zip it with the ground truth label since flatMap may change partitioners.
